validator/src/bootstrap.rs

```diff
@@ -616,629 +500,297 @@ mod without_incremental_snapshots {
-                let mut eligible_rpc_peers = vec![];
-
-                for rpc_peer in rpc_peers.iter() {
-                    cluster_info.get_snapshot_hash_for_node(&rpc_peer.id, |snapshot_hashes| {
-                        for snapshot_hash in snapshot_hashes {
-                            if let Some(ref known_snapshot_hashes) = known_snapshot_hashes {
-                                if !known_snapshot_hashes.contains(snapshot_hash) {
-                                    // Ignore all unknown snapshot hashes
-                                    continue;
-                                }
-                            }
-
-                            if highest_snapshot_hash.is_none()
-                                || snapshot_hash.0 > highest_snapshot_hash.unwrap().0
-                            {
-                                // Found a higher snapshot, remove all nodes with a lower snapshot
-                                eligible_rpc_peers.clear();
-                                highest_snapshot_hash = Some(*snapshot_hash)
-                            }
-
-                            if Some(*snapshot_hash) == highest_snapshot_hash {
-                                eligible_rpc_peers.push(rpc_peer.clone());
-                            }
-                        }
-                    });
-                }
+                let random_peer = &rpc_peers[thread_rng().gen_range(0, rpc_peers.len())];
+                return Some(GetRpcNodeResult {
+                    rpc_contact_info: random_peer.clone(),
+                    snapshot_hash: None,
+                });
+            }
+        }
 
-                match highest_snapshot_hash {
-                    None => {
-                        assert!(eligible_rpc_peers.is_empty());
-                    }
-                    Some(highest_snapshot_hash) => {
-                        if eligible_rpc_peers.is_empty() {
-                            match newer_cluster_snapshot_timeout {
-                                None => newer_cluster_snapshot_timeout = Some(Instant::now()),
-                                Some(newer_cluster_snapshot_timeout) => {
-                                    if newer_cluster_snapshot_timeout.elapsed().as_secs() > 180 {
-                                        warn!("giving up newer snapshot from the cluster");
-                                        return None;
-                                    }
-                                }
-                            }
-                            retry_reason = Some(format!(
-                                "Wait for newer snapshot than local: {:?}",
-                                highest_snapshot_hash
-                            ));
-                            continue;
-                        }
+        let peer_snapshot_hashes = get_peer_snapshot_hashes(
+            cluster_info,
+            validator_config,
+            &rpc_peers,
+            bootstrap_config.incremental_snapshot_fetch,
```

Comment:
Minor/Preference: One of the things I did when making the `mod with_incremental_snapshots()` was to pass in the larger configurations structs to the higher level functions, instead of each individual flag. My preference would be to pass `bootstrap_config` into `get_peer_snapshot_hashes()` (after `validator_config`), instead of the individual flag.

Basically, if the fn already has a parameter for `ValidatorConfig`, then pass in `BootstrapConfig`.

Reply:
I agree with you, when function calls are between modules. And that other module is black box for us. But this local function which makes small work. I think this function used many fields from validator_config early, but now it uses only one field - known_validators. May be better to change validator_config to known_validators?
I think it will be easier to read. At least for me, it is. Especially when you move from functions of higher level abstractions to lower ones. Reading just a function call and seeing the specific parameters, I can already guess what it will do and how. And when there is a large structure instead of parameters, then this is sometimes very difficult to do. (I usually need to jump to structure definition or/and drill down to function implementation to check which fields it uses. )
Look to 7adc650daa88773ef7ababc7b25dba990033193d.
If you don’t like I will make it with bootstrap_config as you proposed.

Reply:
I agree with you, when function calls are between modules. And that other module is black box for us. But this local function which makes small work. I think this function used many fields from validator_config early, but now it uses only one field - known_validators. May be better to change validator_config to known_validators?
I think it will be easier to read. At least for me, it is. Especially when you move from functions of higher level abstractions to lower ones. Reading just a function call and seeing the specific parameters, I can already guess what it will do and how. And when there is a large structure instead of parameters, then this is sometimes very difficult to do. (I usually need to jump to structure definition or/and drill down to function implementation to check which fields it uses. )
Look to 7adc650daa88773ef7ababc7b25dba990033193d.
If you don’t like I will make it with bootstrap_config as you proposed.

Reply:
I mostly try to avoid boolean parameters, because (1) it can be non-obvious the intent when reading call sites that use boolean literals, and (2) it's possible to put the wrong boolean expression as the parameter due to the lack of type safety.

With that said, I'm OK with your version (even if it's not my preference 😊).

A follow-up PR for https://github.com/solana-labs/solana/commit/7adc650daa88773ef7ababc7b25dba990033193d seems fine too.